docs/design/features/objc-interop.md

```diff
@@ -0,0 +1,164 @@
+
+An Objective-C Block variable defined as:
+
+```objective-c
+int(^blk)(int) = ^(int a) { return a*2; };
+```
+
+Can be queried, following the ABI, for its invoke function pointer and dispatched in C as:
+
+```C
+void* fptr = extract_using_abi(blk);
+int b = ((int(*)(id,int))fptr)(blk, a);
+```
+
+**Method dispatch**
+
+The Objective-C [message dispatch mechanism](#message_disp) will require special handling if either exception propagation or variadic variable signatures are desired. If neither exceptions nor variadic arguments are a concern then acquiring the native function pointer to the appropriate message function would be sufficient for internal uses - C# function pointers make this very easy. However, as officially described [here](https://docs.microsoft.com/xamarin/ios/internals/objective-c-selectors), users are permitted to declare their own P/Invoke signatures and perform manual message dispatch - this complicates things.
+
+Xamarin presently defines a set of hand-written assembly for the messaging functions and they handle execptions and most variadic argument cases. Then, utilizing Mono's [`mono_dllmap_insert`](http://docs.go-mono.com/?link=root:/embed) API, maps these entries to override the official Objective-C runtime APIs when loaded through a P/Invoke. This results in users only needing to know the official Objective-C runtime APIs but getting all the benefits of the hand-written assembly.
+
+There are several possible approaches to address this going forward:
+
+* Support a `mono_dllmap_insert` type API to permit overloading.
+* Hardcode these special cases in the runtime with the assembly code.
+* Hardcode these special cases in the runtime but permit the "overrides" to be provided through a managed API.
+
+The last option seems to have the most flexibility and follows the "pay-for-play" principle since some users may not care about exception handling or variadic argument support.
+
+## Application Model
+
+The application model that has been defined by the existing Xamarin scenarios contains multiple layers for consideration. Xamarin applications begin as a C/Objective-C based binary. From there the Mono runtime is activated, types registered with the Objective-C runtime, and the .NET "main" entered.
+
+### Hosting <a name="hosting"></a>
+
+For Xamarin, activation of the .NET Platform is done through Mono's Embedding API and as such is presently exclusive to the Mono version of the .NET Platform. During the Xamarin start-up the Objective-C runtime is implicitly activated since the entry binary is itself an Objective-C application. This entry point concept poses a unique challenge for the CoreCLR given most main stream scenarios leverage the .NET supplied [AppHost](https://docs.microsoft.com/dotnet/core/project-sdk/msbuild-props#useapphost) entry binary.
```

Comment:
What does start up of the Objective-C runtime actually do?

Reply:
I don't actually know what happens during the Objective-C runtime startup, and any requirements related to it, or if anything actually needs to be done for it to happen at all (i.e. it could be done automatically by a library initializer in libobjc.dylib).

Reply:
Right, that's what I thought. I would expect that the startup of the Objective-C runtime happens one way or the other before Objective-C gets used for first time. It is not particularly important that it happens as side-effect of the main app statically linking to it in the Xamarin case.